File: lsp_test/cpp/src/utils/Logger.cpp
#include "utils/Logger.hpp"
#include <iostream>

Logger* Logger::s_instance = nullptr;

Logger::Logger() : m_minLevel(LogLevel::DEBUG) {
    m_logFile.open("app.log", std::ios::app);
}

Logger& Logger::getInstance() {
    if (!s_instance) {
        s_instance = new Logger();
    }
    return *s_instance;
}

void Logger::setMinLevel(LogLevel level) {
    m_minLevel = level;
}

void Logger::log(LogLevel level, const std::string& message) {
    if (level < m_minLevel) return;

    std::string levelStr;
    switch (level) {
    case LogLevel::DEBUG:
        levelStr = "[DEBUG]";
        break;
    case LogLevel::INFO:
        levelStr = "[INFO]";
        break;
    case LogLevel::WARNING:
        levelStr = "[WARNING]";
        break;
    case LogLevel::ERROR:
        levelStr = "[ERROR]";
        break;
    }

    std::string logMessage = levelStr + " " + message;
    std::cout << logMessage << std::endl;
    if (m_logFile.is_open()) {
        m_logFile << logMessage << std::endl;
    }
}

void Logger::debug(const std::string& message) {
    log(LogLevel::DEBUG, message);
}

void Logger::info(const std::string& message) {
    log(LogLevel::INFO, message);
}

void Logger::warning(const std::string& message) {
    log(LogLevel::WARNING, message);
}

void Logger::error(const std::string& message) {
    log(LogLevel::ERROR, message);
}

File: lsp_test/python/utils/logger.py
"""Logging utilities with multiple output levels."""

from enum import IntEnum
from datetime import datetime
from typing import TextIO
import sys


class LogLevel(IntEnum):
    """Log severity levels."""

    DEBUG = 0
    INFO = 1
    WARNING = 2
    ERROR = 3
    CRITICAL = 4


class Logger:
    """Simple logger with level filtering and formatting.

    Test LSP:
    - K on LogLevel should show enum docs
    - gd on LogLevel.WARNING should jump to enum
    """

    _instance: "Logger | None" = None

    def __init__(
        self,
        name: str = "app",
        level: LogLevel = LogLevel.INFO,
        output: TextIO | None = None,
    ) -> None:
        self.name = name
        self.level = level
        self.output = output or sys.stdout
        self._format = "[{timestamp}] [{level}] {name}: {message}"

    @classmethod
    def get_instance(cls) -> "Logger":
        """Get singleton logger instance."""
        if cls._instance is None:
            cls._instance = Logger()
        return cls._instance

    @classmethod
    def set_instance(cls, logger: "Logger") -> None:
        """Set the singleton logger instance."""
        cls._instance = logger

    def _format_message(self, level: LogLevel, message: str) -> str:
        """Format a log message."""
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        return self._format.format(
            timestamp=timestamp,
            level=level.name,
            name=self.name,
            message=message,
        )

    def _log(self, level: LogLevel, message: str) -> None:
        """Internal log method."""
        if level >= self.level:
            formatted = self._format_message(level, message)
            print(formatted, file=self.output)

    def debug(self, message: str) -> None:
        """Log debug message."""
        self._log(LogLevel.DEBUG, message)

    def info(self, message: str) -> None:
        """Log info message."""
        self._log(LogLevel.INFO, message)

    def warning(self, message: str) -> None:
        """Log warning message."""
        self._log(LogLevel.WARNING, message)

    def error(self, message: str) -> None:
        """Log error message."""
        self._log(LogLevel.ERROR, message)

    def critical(self, message: str) -> None:
        """Log critical message."""
        self._log(LogLevel.CRITICAL, message)

    def set_level(self, level: LogLevel) -> None:
        """Change the minimum log level."""
        self.level = level

    def set_format(self, format_string: str) -> None:
        """Set custom format string.

        Available placeholders: {timestamp}, {level}, {name}, {message}
        """
        self._format = format_string
